### PR TITLE
fix: Python 3.8 dropped and support for 3.12 added

### LINKED ISSUE #460: Update edx-cookie-cutter
Update edx-cookie-cutter

- Deprecate [xblock-utils](https://github.com/openedx/edx-cookiecutters/blob/master/cookiecutter-xblock/%7B%7Bcookiecutter.repo_name%7D%7D/requirements/base.in#L8) package
- Also need to update Python & Django versions and Github Actions are also outdated.

### PR DESCRIPTION
**Pull request by CodeWithEmad**
This will update all the actions as follows:

- checkout: v4
- setup-python: v5
- python: 3.12 

Also, Python 3.12 added everywhere.

Close #460 #485

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @CodeWithEmad!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.




#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?



This repository is currently unmaintained.

To get help with finding a technical reviewer, tag the community contributions project manager for this PR in a comment and let them know that your changes are ready for review:

1. On the right-hand side of the PR, find the Contributions project, click the caret in the top right corner to expand it, and check the "Primary PM" field for the name of your PM.
2. Find their GitHub handle [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3548807177/Community+Contributions+Project+Manager#Current-OSPR-Project-Managers).


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

**Comment by itsjeyd** [comment]
Hey @CodeWithEmad, thank you for this contribution!

Let us know when it's ready for review.

**Comment by CodeWithEmad** [comment]
For sure, @itsjeyd. 
I will worl on it in the coming days. In the meanwhile, I wanted to ask about the current state of tests. Should we have the CI actions run only on Python 3.12 or we have to have 3.8 also?

**Comment by itsjeyd** [comment]
@CodeWithEmad Sounds good!

> In the meanwhile, I wanted to ask about the current state of tests. Should we have the CI actions run only on Python 3.12 or we have to have 3.8 also?

I'll forward this question to @feanil.

**Comment by feanil** [comment]
We should continue to run the testing with both Python 3.11 and 3.12 but dropping 3.8 is totally fine across all `openedx` projects.

**Comment by CodeWithEmad** [comment]
@feanil  I'm working on dropping Python 3.8 here, but there are a couple of blockers here:
 the dev image is using the `xblock-sdk` image,

https://github.com/openedx/edx-cookiecutters/blob/46948ba89cdbbd29c90f59ebf405cd867bfa05b3/cookiecutter-xblock/%7B%7Bcookiecutter.repo_name%7D%7D/Dockerfile#L2

which is still using Python 3.8, https://github.com/openedx/xblock-sdk/blob/1760448ebb073edba35be87d99f474d90d9421ac/Dockerfile#L19
I found this https://github.com/openedx/xblock-sdk/pull/350, but it's stuck for a while. 
Also, lint jobs are failing since there's no pkg_resources and edx-lint is still using it: https://github.com/openedx/edx-lint/pull/398

https://github.com/openedx/edx-cookiecutters/actions/runs/10206678813/job/28240034237?pr=471

**Comment by feanil** [comment]
@CodeWithEmad I'll see if we can land the xblock-sdk issue first then we can update that as a part of the PR here.

**Comment by farhan** [comment]
Issue will be fixed in https://github.com/openedx/xblock-sdk/pull/397
Hopefully this PR will be unblocked to work on once its merged and docker image pushed.
FYI: @CodeWithEmad

**Comment by CodeWithEmad** [comment]
Thanks, @farhan. I'll continue this in the coming weeks.

File: .github/workflows/ci.yml
Comment (farhan) [inline .github/workflows/ci.yml:10]: Why are we removing them?

File: cookiecutter-xblock/{{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (farhan) [inline cookiecutter-xblock/{{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py:26]: We are calling the `xblock/utils/resources.py` methods where application and we can use them here. Please study below PR to follow it

https://github.com/openedx/xblock-in-video-quiz/pull/212/files#diff-cf5d4b07c1f2d28759f30b933fbdacaebf074f45fe08f1890b1fec6a9a049148R31

**Review by farhan — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml
Comment (CodeWithEmad) [reply .github/workflows/ci.yml:10]: The workflow_dispatch trigger does not require or support a branches key. By default, it allows manual triggering regardless of branch.
https://docs.github.com/en/webhooks/webhook-events-and-payloads#workflow_dispatch

File: cookiecutter-xblock/{{cookiecutter.repo_name}}/requirements/base.in
Comment (CodeWithEmad) [reply cookiecutter-xblock/{{cookiecutter.repo_name}}/requirements/base.in:8]: You're right. I'll remove it.

File: cookiecutter-xblock/{{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (CodeWithEmad) [reply cookiecutter-xblock/{{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: Nice.

File: cookiecutter-xblock/{{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (CodeWithEmad) [reply cookiecutter-xblock/{{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py:103]: I guess you're comparing this with `invideoquiz xblock`, @farhan.
There's no public directory here. unlike `invideoquiz`, all the static files are stored inside `static` directory.

**Comment by CodeWithEmad** [comment]
@farhan I created an XBlock from the cookiecutter and installed it in the xblock-sdk as well. 

image

It's working as expected.

**Review by farhan — APPROVED** [review]
Good work ⭐

**Comment by itsjeyd** [comment]
@farhan @CodeWithEmad Looks like this PR is ready to merge! 🚀

**Comment by farhan** [comment]
I don't have merging rights
@CodeWithEmad Do you have merging rights?
or @kdmccormick @feanil can help

File: cookiecutter-django-app/README.rst
Comment (irtazaakram) [inline cookiecutter-django-app/README.rst:57]: We should bump pip version to `23.2`

File: cookiecutter-xblock/{{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (irtazaakram) [inline cookiecutter-xblock/{{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py:14]: I don't think we need to support `xblockutils` in cookiecutter. All new projects should be using `xblock.utils`

**Review by irtazaakram — CHANGES_REQUESTED** [review]
Just a few minor tweaks, but overall, it looks great!

File: cookiecutter-xblock/{{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py
Comment (farhan) [reply cookiecutter-xblock/{{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py:14]: +1 , we don't need to provide backward support in cookie cutter

Comment (CodeWithEmad) [reply cookiecutter-xblock/{{cookiecutter.repo_name}}/{{cookiecutter.package_name}}/{{cookiecutter.package_name}}.py]: Sure. I'll drop the backward compatibility support.

**Review by irtazaakram — APPROVED** [review]

**Review by farhan — APPROVED** [review]

**Comment by CodeWithEmad** [comment]
@itsjeyd Looks like this is ready to merge!

**Review by kdmccormick — APPROVED** [review]
I have not reviewed line-by-line, but I trust the Aximprovement team's reviews here.

@feanil , for context, Emad and I spoke about whether we need to put 3.11 testing back in place for the cookie-cut result repo. My sense was that we are so close to being on 3.12 project-wide that it would be a waste of effort and CI cycles to test both versions for new packages, especially since the services running 3.11 will transitively test much of the package code via its own 3.11 CI checks. If you disagree LMK and we can put back 3.11 testing in a quick followup.

**Comment by kdmccormick** [comment]
Thanks @CodeWithEmad !

**Pull request merged**
